### PR TITLE
WIP: Merge System.Threading.Thread into Internal.Runtime.Augments.RuntimeThread.

### PR DESCRIPTION
**Pull request by filipnavara**
This is just a proof of concept. It will likely break VS Debugger in its current shape, but it depends on how the VS Debugger accesses the `_managedThreadId` field.

Here's a background on what I am trying to solve. It was already discussed in https://github.com/mono/mono/pull/12470#discussion_r248865698, but I will restate it here for clarity.

Right now CoreFX implements `System.Threading.Thread` using calls to `Internal.Runtime.Augments.RuntimeThread` in CoreLib. The CoreRT implementation is quite clean, but for historical reasons the CoreCLR implementation is not. It's split into `RuntimeThread` and another internal `System.Threading.Thread` type in CoreLib. This is an unfortunate situation where two `System.Threading.Thread` types exist, with different, but overlapping, functionality, in two different assemblies. To add insult to the injury some types in CoreLib reference the `System.Threading.Thread` directly instead of `RuntimeThread`, which makes code sharing with CoreRT/Mono difficult.

There are three proposed solutions for a cleanup:

1) Get rid of `System.Threading.Thread` in CoreLib and fold its functionality into `RuntimeThread`. This is implemented by this PR. Additionally, `RuntimeThread` namespace should be changed to Internal.Threading.RuntimeThread to make it match the convention used for most types like that (not done in this PR and will require changes to CoreRT and CoreFX too).

2) Keep  `System.Threading.Thread` in CoreLib, but ensure that it is called only from `RuntimeThread` and unmanaged code. All other code would reference `RuntimeThread` only.

3) Move `System.Threading.Thread` from CoreFX back to CoreLib, including its transitive dependency closure and get rid of `RuntimeThread`.

Option 2) seems to be least favorable. The only thing going for it is that it won't break VS debugger and it's very limited in scope to the CoreCLR repository. Ultimately it doesn't solve the problem with two classes with the same name though.

Option 3) will require most shuffling of code between repositories.

### CODE REVIEW

**Comment by jkotas** [comment]
cc @marek-safar

File: src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs [hunk 1]
```diff
@@ -14,22 +14,135 @@ namespace Internal.Runtime.Augments
 {
     public class RuntimeThread : CriticalFinalizerObject
     {
+        /*=========================================================================
```
Comment (marek-safar) [inline src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs]: Hmm, there is more code duplicating with CoreFX Thread that I expected

Comment (jkotas) [reply src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs]: Do you have anything specific you have in mind?

**Comment by jkotas** [comment]
Let's wait for the Environment CoreFX->CoreCLR move to finish first, and then decide which option to take for this one.

**Comment by filipnavara** [comment]
Sure. This is low-priority one, but I decided to tackle it because it bit me couple of times that the class name resolution didn't work as I expected (ie. `using Thread = ...` was effectively ignored because the class was in the `System.Threading` namespace). I'll be happy to help with a solution (whichever one you decide on) once the work on `Environment` is finished.

**Comment by filipnavara** [comment]
Any new thoughts about this?

**Comment by marek-safar** [comment]
@jkotas @stephentoub could you please advice Filip on future steps

**Comment by jkotas** [comment]
I think we should bring the public `System.Threading.Thread` with its transitive closure back to CoreLib. @stephentoub @marek-safar Do you agree?

If we go with this plan, the first step should be to identify all files in CoreFX that need to be moved to CoreLib and move them to CoreLib shared partition in the CoreFX repo. (Except Thread.cs itself since it will require refactoring to integrate back to CoreLib.)

**Comment by stephentoub** [comment]
> I think we should bring the public System.Threading.Thread with its transitive closure back to CoreLib. @stephentoub @marek-safar Do you agree?

In theory, yes. It's be good to see the full transitive list before pulling the trigger. From a quick skim, it looks like the worst of it would be some permissions stuff via AppDomain, which should probably be in Corelib, anyway.

**Comment by marek-safar** [comment]
> I think we should bring the public `System.Threading.Thread` with its transitive closure back to CoreLib.

This makes sense, all VM specific types are in CoreLib and this will hopefully make few dependencies simpler as well. I didn't check Thread dependencies chain but it's hopefully not too bad.

**Comment by filipnavara** [comment]
Here's the list:

- AppDomain.cs
- AppDomainSetup.cs
- ArrayList.cs
- CompressedStack.cs
- IIdentity.cs
- IPermission.cs
- IPrincipal.cs
- ISecurityEncodable.cs
- IStackWalk.cs
- LocalDataStoreSlot.cs
- PermissionSet.cs
- PermissionState.cs
- PrincipalPolicy.cs
- SecurityElement.cs
- Thread.cs
- Thread.Unix.cs
- Thread.Windows.cs

The most troubling one is `ArrayList` pulled by `SecurityElement`.

**Comment by filipnavara** [comment]
I tried copying the above files into CoreRT's System.Private.CoreLib to see if it can compile. I've got to the point where only resource strings are missing, so the list above should be fairly complete.

**Comment by stephentoub** [comment]
Thanks for the list.

> System.Runtime.Extensions\src\System\Collections\ArrayList.cs

Is this actually exposed through API, or could the implementation just use `List<object>`?

**Comment by filipnavara** [comment]
> Is this actually exposed through API, or could the implementation just use `List<object>`?

Unfortunately it is exposed. It's the biggest dependency of them all.

**Comment by jkotas** [comment]
`ArrayList` is old sibling of `Hashtable`. I do not feel bad about having it in CoreLib.

**Comment by filipnavara** [comment]
I'm going to close this PR, but feel free to continue discussion.

Is anyone planning to work on it? I'm notoriously bad at multi-repository code transfers, but if you are willing to bear with me I can try to reproduce the same PR sequence as for the `Environment` transfer.

**Comment by jkotas** [comment]
We will be glad if you can help.

**Comment by stephentoub** [comment]
And if you need help and get stuck, please let me know; I'll be happy to.

**Comment by filipnavara** [comment]
Alright. I will get to work later today.

**Pull request closed without merging**